### ISSUE TITLE
Usage of `Self` in impl blocks

### ISSUE THREAD

**Issue by 54k1** [tuqqu/oxide-lang#2]
I tried to use Self in a method in the following manner and could not get it to work
```rust
struct Circle {
    radius: float
}

impl Circle {
    fn larger_circle() -> Self {
        return Self {
            radius: self.radius()+10
        }
    }
}
```

I think there are two issues here.
1. `Self` cannot be used as return type.
2. `Self` cannot be used in call expressions.

Solutions:
1. Add a new variable to parser to keep track of the struct which the current impl block points to. We could call it `current_struct_decl` maybe, and it can be an optional. It must be set whenever we come across an impl block, and reset once out of the impl block.
```rust
struct Circle {...}
impl Circle {
// set current_struct_decl to Circle
}
// set current_struct_decl to None
```
2. Update [consume_type](https://github.com/tuqqu/oxide-lang/blob/48f3e1d147d1e831a3e7b79c67fce849d4f03280/src/parser/mod.rs#L1208) to support `Self`. We need to add a check for `Self` and return the appropriate `struct_decl` in case we are in an impl block. If we are not in an impl block, then `Self` does not make sense, and we need to raise an error.
3. In [call_expr](https://github.com/tuqqu/oxide-lang/blob/48f3e1d147d1e831a3e7b79c67fce849d4f03280/src/parser/mod.rs#L895) , currently, only the lexeme is being checked if it is present in the structs. `impl` has the lexeme "impl", and therefore is not matched in the structs array. A match for `Self` also must be checked in this case.

Please let me know if I'm missing anything in these issues and their proposed solutions. I can take this issue up once you approve. Very cool project btw! :)

**Comment by tuqqu** [comment]
Hello! Thanks for taking the time to test and improve things!
Yes, as of now `Self` usage is limited  to `Self::` call expressions only (similar to `self.` but for statics).

The solution you came up with is a very reasonable one, I am amazed at how well you spotted the exact locations to make improvements.

1. That's a very good approach. At first I would have thought that it should be done in the interpreter instead, but I think you're actually right, there is no reason to involve interpreter when all can be resolved in the parser. Frankly, I might reconsider my current implementation of `Self::` resolving, once your approach is implemented :)
2. Correct.
3. Correct.

I don't think you missed anything to be honest and I would be very glad to see contributions.

**Comment by tuqqu** [comment]
Implemented in #3
